src/sage/rings/generic.py

```diff
@@ -196,6 +202,55 @@ def remainders(self, x):
             X = [X[i // 2] % V[i] for i in range(len(V))]
         return X
 
+    _crt_bases = None
+
+    def interpolation(self, xs):
+        r"""
+        Given a sequence ``xs`` of values, one per leaf, return a
+        single element `x` which is congruent to the `i`\th value in
+        ``xs`` modulo the `i`\th leaf, for all `i`.
+
+        This is an explicit version of the Chinese remainder theorem;
+        see also :meth:`CRT`. Using this product tree is faster for
+        repeated calls since the required CRT bases are cached after
+        the first run.
+
+        The base ring must support the :func:`xgcd` function for this
+        method to work.
+
+        EXAMPLES::
+
+            sage: from sage.rings.generic import ProductTree
+            sage: vs = prime_range(100)
+            sage: tree = ProductTree(vs)
+            sage: tree.interpolation([1, 1, 2, 1, 9, 1, 7, 15, 8, 20, 15, 6, 27, 11, 2, 6, 0, 25, 49, 5, 51, 4, 19, 74, 13])
+            1085749272377676749812331719267
+
+        This method is faster than :func:`CRT` for repeated calls with
+        the same moduli::
+
+            sage: vs = prime_range(1000,2000)
+            sage: rs = list(range(len(vs)))
+            sage: tree = ProductTree(vs)
+            sage: %timeit CRT(rs,vs)                # not tested
+            324 µs ± 637 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
+            sage: %timeit tree.interpolation(rs)    # not tested
+            102 µs ± 92.5 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
+        """
```

Comment:
I am not sure if this example shows your point. Perhaps you want to change `rs` for the later runs and show that `interpolation` runs faster than the first one? Like the following,

```python
sage: vs = prime_range(100)
sage: tree = ProductTree(vs)
sage: rs = [randint(1,100) for _ in range(len(vs))]
sage: time CRT(rs,vs)
CPU times: user 608 µs, sys: 1 µs, total: 609 µs
Wall time: 618 µs
1675616340214929388822501993863068055
sage: time tree.interpolation(rs)
CPU times: user 501 µs, sys: 1 µs, total: 502 µs
Wall time: 511 µs
1675616340214929388822501993863068055
sage: rs = [randint(1,100) for _ in range(len(vs))]
sage: time CRT(rs,vs)
CPU times: user 289 µs, sys: 0 ns, total: 289 µs
Wall time: 293 µs
751801689301901428485846165281945864
sage: time tree.interpolation(rs)
CPU times: user 123 µs, sys: 0 ns, total: 123 µs
Wall time: 127 µs
751801689301901428485846165281945864
sage: rs = [randint(1,100) for _ in range(len(vs))]
sage: time CRT(rs,vs)
CPU times: user 574 µs, sys: 1 µs, total: 575 µs
Wall time: 584 µs
2019633189310335097889921080371891959
sage: time tree.interpolation(rs)
CPU times: user 218 µs, sys: 18 µs, total: 236 µs
Wall time: 318 µs
2019633189310335097889921080371891959
sage: rs = [randint(1,100) for _ in range(len(vs))]
sage: time CRT(rs,vs)
CPU times: user 487 µs, sys: 1e+03 ns, total: 488 µs
Wall time: 495 µs
351522767900743874314752778753118403
sage: time tree.interpolation(rs)
CPU times: user 231 µs, sys: 1e+03 ns, total: 232 µs
Wall time: 239 µs
351522767900743874314752778753118403
```